node/src/components/linear_chain.rs

```diff
@@ -286,60 +335,135 @@ where
+                        }
+                    })
+            }
+            .event(|(maybe_block, fs, is_bonded)| Event::IsBonded(maybe_block, fs, is_bonded)),
+            Event::IsBonded(maybe_block, fs, is_bonded) => {
+                match maybe_block {
+                    Some(block) if is_bonded => {
+                        // Known block and signature from a bonded validator.
+                        self.add_finality_signature(*fs);
+                        let old_count = block.proofs().len();
+                        let (block, mut effects) =
+                            self.add_pending_finality_signatures(*block, effect_builder);
+                        if block.proofs().len() > old_count {
+                            let block = Box::new(block);
+                            effects.extend(effect_builder.put_block_to_storage(block).ignore());
+                        }
+                        effects
+                    }
+                    Some(_block) if !is_bonded => {
+                        // Known block but not a bonded validator (neither in current nor in future
+                        // era). Validator cannot be from the future era b/c
+                        // we wouldn't have the block.
+                        let FinalitySignature {
+                            public_key,
+                            block_hash,
+                            ..
+                        } = *fs;
+                        warn!(
+                            validator = %public_key,
+                            %block_hash,
+                            "Received a signature from a validator that is not bonded."
+                        );
+                        // TODO: Disconnect from the sender.
+                        Effects::new()
+                    }
+                    None if is_bonded => {
+                        // Unknown block but validator is bonded.
+                        // We should finalize the same block eventually. Either in this or in the
+                        // next era.
+                        self.add_finality_signature(*fs);
+                        Effects::new()
+                    }
+                    None if !is_bonded => {
+                        // Unknown block from unknown validator.
+                        // We don't have the block in storage and validator is not bonded in this
+                        // era or in the next one.
+                        let FinalitySignature {
+                            public_key,
+                            block_hash,
+                            ..
+                        } = *fs;
+                        warn!(
+                            validator = %public_key,
+                            %block_hash,
+                            "Received a signature from a validator that is not bonded."
+                        );
+                        // TODO: Disconnect from the sender.
+                        Effects::new()
```

Comment:
Isn't this identical to the other `!is_bonded` match arm, apart from the comment?
I'd avoid nesting `match`es here anyway:
```rust
Event::IsBonded(maybe_block, fs, false) => { /* this branch */ },
Event::IsBonded(Some(block), fs, true) => …,
Event::IsBonded(None, fs, true) => …,
```

Reply:
It is identical but I thought it's important to explain why we do what we do – hence different cases and comments. I will see if I can unify them in some readable manner.

Reply:
Done in [`c0bd13f` (#683)](https://github.com/CasperLabs/casper-node/pull/683/commits/c0bd13f70e41208382667535631dd62c9a381c5c)